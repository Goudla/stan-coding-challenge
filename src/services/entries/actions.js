/*
 * Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, payload: { var: var } }
 *    }
 */

import type { Entry } from './types';

import {
  LOAD_PHOTOS_REQUEST,
  LOAD_PHOTOS_SUCCESS,
  LOAD_PHOTOS_FAILURE
} from './constants';

/**
 * Load the entries, this action starts the request saga
 * @return {object} An action object with a type of LOAD_PHOTOS_REQUEST
 */
export function loadEntries() {
  return {
    type: LOAD_PHOTOS_REQUEST
  };
}

/**
 * Dispatched when the entries are loaded by the request saga
 * @param  {array} entries The entries data
 * @return {object} An action object with a type of LOAD_PHOTOS_SUCCESS passing the entries
 */
export function entriesLoaded(entries: { [string]: Entry }) {
  return {
    type: LOAD_PHOTOS_SUCCESS,
    payload: {
      entries
    }
  };
}

/**
 * Dispatched when loading the entries fails
 * @param  {object} error The error
 * @return {object} An action object with a type of LOAD_PHOTOS_FAILURE passing the error
 */
export function entriesLoadingFailure(error: Error) {
  return {
    type: LOAD_PHOTOS_FAILURE,
    payload: error,
    error: true
  };
}
