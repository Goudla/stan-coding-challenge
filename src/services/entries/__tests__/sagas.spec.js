import { put, takeLatest } from 'redux-saga/effects';
import keyBy from 'lodash.keyby';

import { LOAD_PHOTOS_REQUEST } from '../constants';
import { entriesLoaded, entriesLoadingFailure } from '../actions';

import entriesData, { getEntries } from '../sagas';

describe('getEntries saga', () => {
  let getEntriesGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getEntriesGenerator = getEntries();

    const callDescriptor = getEntriesGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the entriesLoaded action if it requests the data successfully', () => {
    const response = {
      data: {
        entries: [
          {
            title: 'test',
            description: 'test',
            programType: 'series',
            images: {
              'Poster Art': {
                url: "https://test.com",
                width: 1234,
                height: 1234
              }
            },
            releaseYear: 1234
          }
        ]
      }
    };
    const responseKeyedById = keyBy(response.data.entries, 'title');
    const putDescriptor = getEntriesGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(entriesLoaded(responseKeyedById)));
  });

  it('should call the entriesLoadingFailure action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getEntriesGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(entriesLoadingFailure(response)));
  });
});

describe('entriesData saga', () => {
  const entriesDataSaga = entriesData();

  it('should start task to watch for LOAD_PHOTOS_REQUEST action', () => {
    const takeLatestDescriptor = entriesDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(
      takeLatest(LOAD_PHOTOS_REQUEST, getEntries)
    );
  });
});
