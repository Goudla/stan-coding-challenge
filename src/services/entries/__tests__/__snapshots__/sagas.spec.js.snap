// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getEntries saga should call the entriesLoadingFailure action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "/feed/sample.json",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getEntries saga should dispatch the entriesLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "/feed/sample.json",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
